cmd/authgear/importer/importer.go

```diff
@@ -0,0 +1,189 @@
+	r := csv.NewReader(f)
+	for {
+		record, err := r.Read()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		if !headerSkipped {
+			headerSkipped = true
+			continue
+		}
+
+		userID := record[0]
+		rawEmail := record[1]
+		name := record[2]
+		passwordHash := record[3]
+
+		stdAttrs, err := StandardAttributes(name)
+		if err != nil {
+			return err
+		}
+
+		emailChecker := loginid.EmailChecker{
+			Config: i.EmailConfig,
+		}
+		validationCtx := &validation.Context{}
+		emailChecker.Validate(validationCtx, rawEmail)
+		err = validationCtx.Error(fmt.Sprintf("invalid email: %v", rawEmail))
+		if err != nil {
+			return err
+		}
+
+		emailNormalizer := loginid.EmailNormalizer{
+			Config: i.EmailConfig,
+		}
+
+		loginID, err := emailNormalizer.Normalize(rawEmail)
+		if err != nil {
+			return err
+		}
+
+		uniqueKey, err := emailNormalizer.ComputeUniqueKey(loginID)
+		if err != nil {
+			return err
+		}
+
+		claims, err := Claims(loginID)
+		if err != nil {
+			return err
+		}
+
+		insertStmts := []db.InsertBuilder{
+			i.SQLBuilderApp.Insert(i.SQLBuilderApp.TableName("_auth_user")).Columns(
+				"id",
+				"created_at",
+				"updated_at",
+				"is_disabled",
+				"standard_attributes",
+			).Values(
+				userID,
+				now,
+				now,
+				false,
+				stdAttrs,
+			).Suffix("ON CONFLICT (id) DO UPDATE SET standard_attributes = EXCLUDED.standard_attributes"),
+			i.SQLBuilderApp.Insert(i.SQLBuilderApp.TableName("_auth_identity")).Columns(
+				"id",
+				"type",
+				"user_id",
+				"created_at",
+				"updated_at",
+			).Values(
+				userID,
+				"login_id",
+				userID,
+				now,
+				now,
+			).Suffix("ON CONFLICT DO NOTHING"),
+			i.SQLBuilderApp.Insert(i.SQLBuilderApp.TableName("_auth_identity_login_id")).Columns(
+				"id",
+				"login_id_key",
+				"login_id",
+				"claims",
+				"original_login_id",
+				"unique_key",
+				"login_id_type",
+			).Values(
+				userID,
+				"email",
+				loginID,
+				claims,
+				rawEmail,
+				uniqueKey,
+				"email",
+			).Suffix("ON CONFLICT (id) DO UPDATE SET login_id = EXCLUDED.login_id, claims = EXCLUDED.claims, original_login_id = EXCLUDED.original_login_id, unique_key = EXCLUDED.unique_key"),
+			i.SQLBuilderApp.Insert(i.SQLBuilderApp.TableName("_auth_authenticator")).Columns(
+				"id",
+				"type",
+				"user_id",
+				"created_at",
+				"updated_at",
+				"is_default",
+				"kind",
+			).Values(
+				userID,
+				"password",
+				userID,
+				now,
+				now,
+				true,
+				"primary",
+			).Suffix("ON CONFLICT DO NOTHING"),
+			i.SQLBuilderApp.Insert(i.SQLBuilderApp.TableName("_auth_authenticator_password")).Columns(
+				"id",
+				"password_hash",
+			).Values(
+				userID,
+				passwordHash,
+			).Suffix("ON CONFLICT (id) DO UPDATE SET password_hash = EXCLUDED.password_hash"),
+		}
+
+		err = i.Handle.WithTx(func() (err error) {
+			for _, stmt := range insertStmts {
+				_, err = i.SQLExecutor.ExecWith(stmt)
+				if err != nil {
+					return
```

Comment:
We return nil here instead of error, is it because we want to skip the problematic row rather than stop the whole import?

It would be better if the importer can print out a simple import summary like number of successful import, the problematic row number? So that when there are any problems during import, we can spot the problematic data easily.

Reply:
We do return error here. `err` is declared in the function signature.

Reply:
A transaction is opened per row in the CSV. The program stops as soon as it encounter an error. I agree that we should skip problematic row. I suggest we print the problematic record to stderr, and also print a summary to the stdout at the end of the import.